### PR TITLE
TouchID Attestation [WIP]

### PR DESCRIPTION
**Pull request by ericmarkmartin**
- wip
- feat(touchid-attest): initial implementation of touch id attestation

Implements #33.

Currently missing steps 4 and 5, and the validation of the certificate chain
against Apple's root certificates.

- [x] cargo fmt has been run
- [x] cargo test has been run and passes
- [x] documentation has been updated with relevant examples (if relevant)

### CODE REVIEW

**Comment by Firstyear** [comment]
As a tip, you can use Relates: #issuenumber in yoru commit message to auto link to the issue. You can use 'Fixes #number' to have the issue auto-close when it's merged.

Anyway, The first thing is that this is missing tests, so I think that's needed before we review this further. You'll need to capture a registration from an ipad or iphone to the example server that you can use for this.

As well, the attestation type is "AppleAnonymous" not "Apple".

Thanks,

**Comment by ericmarkmartin** [comment]
> As well, the attestation type is "AppleAnonymous" not "Apple".

Isn't the attestation type "AnonCa"? Do you mean the attestation format?

File: src/attestation.rs [hunk 1]
```diff
@@ -23,6 +23,7 @@ pub(crate) enum AttestationFormat {
     AndroidKey,
     AndroidSafetyNet,
     FIDOU2F,
+    Apple,
```
Comment (Firstyear) [inline src/attestation.rs:26]: This line should be AppleAnonymous

**Comment by Firstyear** [comment]
> > You'll need to capture a registration from an ipad or iphone to the example server that you can use for this.
> 
> I did attempt to do this but I was not prompted in mobile safari to authenticate with touch id.

Yep, you need to likely tweak settings in the example server to achieve this correctly. See `get_attestation_preference` for how to req the full attestation, and you may need to alter `UserVerificationPolicy::Required` depending on the type of device you have. There are also other reasons you may not have been able to auth (iOS version for example).

File: src/attestation.rs (hunk 1 above)
Comment (ericmarkmartin) [reply src/attestation.rs:26]: Got it—this is despite the corresponding fmt string being "Apple"?

**Comment by ericmarkmartin** [comment]
> > > You'll need to capture a registration from an ipad or iphone to the example server that you can use for this.
> > 
> > 
> > I did attempt to do this but I was not prompted in mobile safari to authenticate with touch id.
> 
> Yep, you need to likely tweak settings in the example server to achieve this correctly. See `get_attestation_preference` for how to req the full attestation, and you may need to alter `UserVerificationPolicy::Required` depending on the type of device you have. There are also other reasons you may not have been able to auth (iOS version for example).

I've tried every permutation of `Direct` and `None` for `get_attestation_preference` and `Required` and `Discouraged` for `UserVerificationPolicy`, and still no luck on getting a touch id prompt.

I'm sure it's not an iOS versioning issue because webauthn.io allows me to attest with touch id.

File: src/attestation.rs (hunk 1 above)
Comment (Firstyear) [reply src/attestation.rs:26]: That needs to change too then. The spec calls it AppleAnonymous so it should always be AppleAnonymous in all contexts.

**Comment by Firstyear** [comment]
> > Yep, you need to likely tweak settings in the example server to achieve this correctly. See `get_attestation_preference` for how to req the full attestation, and you may need to alter `UserVerificationPolicy::Required` depending on the type of device you have. There are also other reasons you may not have been able to auth (iOS version for example).
> 
> I've tried every permutation of `Direct` and `None` for `get_attestation_preference` and `Required` and `Discouraged` for `UserVerificationPolicy`, and still no luck on getting a touch id prompt.
> 
> I'm sure it's not an iOS versioning issue because webauthn.io allows me to attest with touch id.

No, there was iOS version issues previously. Saying this provided you are on the latest iOS 14 or iPadOS 14 it should work.

The other consideration is that you must do the challenge over https. If you need you can also debug the client by using safari on a mac and connecting a lightning cable to the iOS device, which allows you access to the safari log of the phone on the mac. 

Safari > Preferences > Advanced > Show Develop Menu, restart safari, then attach the device and go develop -> device name -> select the page you want to inspect

File: src/attestation.rs (hunk 1 above)
Comment (ericmarkmartin) [reply src/attestation.rs:26]: According to the [docs](https://webkit.org/blog/11312/meet-face-id-and-touch-id-for-the-web/)

```

$$attStmtType //= (
                       fmt: "apple",
                       attStmt: appleStmtFormat
                   )
```

it seems like the string should be "apple."

Comment (Firstyear) [reply src/attestation.rs:26]: Ahh that's lovely and confusing as all the other references to it are Apple Anonymous. In that case, you are correct :)

**Comment by ericmarkmartin** [comment]
> > > Yep, you need to likely tweak settings in the example server to achieve this correctly. See `get_attestation_preference` for how to req the full attestation, and you may need to alter `UserVerificationPolicy::Required` depending on the type of device you have. There are also other reasons you may not have been able to auth (iOS version for example).
> > 
> > 
> > I've tried every permutation of `Direct` and `None` for `get_attestation_preference` and `Required` and `Discouraged` for `UserVerificationPolicy`, and still no luck on getting a touch id prompt.
> > I'm sure it's not an iOS versioning issue because webauthn.io allows me to attest with touch id.
> 
> No, there was iOS version issues previously. Saying this provided you are on the latest iOS 14 or iPadOS 14 it should work.

Yeah, I'm running 14, so that shouldn't be the issue.

> The other consideration is that you must do the challenge over https.

Just tried it over https—still no touch id prompt.

> If you need you can also debug the client by using safari on a mac and connecting a lightning cable to the iOS device, which allows you access to the safari log of the phone on the mac.
> 
> Safari > Preferences > Advanced > Show Develop Menu, restart safari, then attach the device and go develop -> device name -> select the page you want to inspect

That's good to know—I might have to try that.

File: src/attestation.rs (hunk 1 above)
Comment (ericmarkmartin) [reply src/attestation.rs:26]: Indeed, not the most straightforward setup I've seen. To clarify, the variant name should still change to `AppleAnonymous`, correct?

**Comment by Firstyear** [comment]
> > If you need you can also debug the client by using safari on a mac and connecting a lightning cable to the iOS device, which allows you access to the safari log of the phone on the mac.
> > Safari > Preferences > Advanced > Show Develop Menu, restart safari, then attach the device and go develop -> device name -> select the page you want to inspect
> 
> That's good to know—I might have to try that.

So I can get it to work. You may want to try something like:

`cargo run --example tide -- -b 0.0.0.0:8443 -i amethyst.local -n amethyst.local -s -d -o https://amethyst.local:8443`

This replaces the id's with amethyst (which is my computer name) and the .local is avahi/bonjour multicast dns that macos/ios support (windows supports with apple bonjour client, linux with avahi). The origin has to match the hostname you connect to else the client may chose not to present any credentials to the server which could be what's going wrong.

You then need to access the demo server with the same url, so `https://amethyst.local:8443` in this example.

Anyway, if you need a test vector, you should be able to transform this into a test case - but I still encourage you to attempt to setup the example server so you can get your own test vector, as it will help you to understand some of these "fun" edge cases in webauthn :) 

```
tide::actors handle Register -> (username: "William", reg: RegisterPublicKeyCredential { id: "kppoCYjQKnXt0IglgtERmdegpTI", raw_id: Base64UrlSafeData([146, 154, 104, 9, 136, 208, 42, 117, 237, 208, 136, 37, 130, 209, 17, 153, 215, 160, 165, 50]), response: AuthenticatorAttestationResponseRaw { attestation_object: Base64UrlSafeData([163, 99, 102, 109, 116, 101, 97, 112, 112, 108, 101, 103, 97, 116, 116, 83, 116, 109, 116, 161, 99, 120, 53, 99, 130, 89, 2, 72, 48, 130, 2, 68, 48, 130, 1, 201, 160, 3, 2, 1, 2, 2, 6, 1, 118, 54, 151, 200, 243, 48, 10, 6, 8, 42, 134, 72, 206, 61, 4, 3, 2, 48, 72, 49, 28, 48, 26, 6, 3, 85, 4, 3, 12, 19, 65, 112, 112, 108, 101, 32, 87, 101, 98, 65, 117, 116, 104, 110, 32, 67, 65, 32, 49, 49, 19, 48, 17, 6, 3, 85, 4, 10, 12, 10, 65, 112, 112, 108, 101, 32, 73, 110, 99, 46, 49, 19, 48, 17, 6, 3, 85, 4, 8, 12, 10, 67, 97, 108, 105, 102, 111, 114, 110, 105, 97, 48, 30, 23, 13, 50, 48, 49, 50, 48, 53, 48, 53, 52, 56, 48, 53, 90, 23, 13, 50, 48, 49, 50, 48, 56, 48, 53, 52, 56, 48, 53, 90, 48, 129, 145, 49, 73, 48, 71, 6, 3, 85, 4, 3, 12, 64, 50, 98, 100, 51, 52, 48, 48, 101, 53, 54, 98, 54, 54, 98, 53, 52, 48, 102, 56, 48, 56, 102, 100, 50, 101, 98, 53, 98, 101, 99, 56, 102, 52, 53, 55, 50, 101, 101, 101, 57, 54, 52, 51, 56, 52, 102, 50, 97, 55, 100, 102, 102, 57, 52, 100, 97, 100, 97, 48, 57, 102, 57, 51, 51, 49, 26, 48, 24, 6, 3, 85, 4, 11, 12, 17, 65, 65, 65, 32, 67, 101, 114, 116, 105, 102, 105, 99, 97, 116, 105, 111, 110, 49, 19, 48, 17, 6, 3, 85, 4, 10, 12, 10, 65, 112, 112, 108, 101, 32, 73, 110, 99, 46, 49, 19, 48, 17, 6, 3, 85, 4, 8, 12, 10, 67, 97, 108, 105, 102, 111, 114, 110, 105, 97, 48, 89, 48, 19, 6, 7, 42, 134, 72, 206, 61, 2, 1, 6, 8, 42, 134, 72, 206, 61, 3, 1, 7, 3, 66, 0, 4, 100, 16, 220, 11, 79, 59, 86, 121, 251, 141, 86, 158, 12, 121, 150, 102, 205, 227, 243, 209, 235, 220, 18, 74, 0, 193, 45, 139, 121, 50, 245, 71, 219, 187, 123, 78, 63, 94, 22, 115, 111, 37, 237, 6, 197, 253, 242, 209, 246, 56, 183, 165, 217, 97, 252, 139, 171, 39, 111, 39, 188, 113, 4, 62, 163, 85, 48, 83, 48, 12, 6, 3, 85, 29, 19, 1, 1, 255, 4, 2, 48, 0, 48, 14, 6, 3, 85, 29, 15, 1, 1, 255, 4, 4, 3, 2, 4, 240, 48, 51, 6, 9, 42, 134, 72, 134, 247, 99, 100, 8, 2, 4, 38, 48, 36, 161, 34, 4, 32, 222, 28, 62, 247, 32, 241, 183, 110, 177, 204, 39, 69, 99, 248, 135, 202, 149, 50, 74, 118, 63, 116, 3, 3, 39, 166, 80, 172, 62, 237, 10, 138, 48, 10, 6, 8, 42, 134, 72, 206, 61, 4, 3, 2, 3, 105, 0, 48, 102, 2, 49, 0, 236, 245, 22, 193, 45, 78, 107, 193, 85, 4, 1, 158, 227, 73, 116, 140, 172, 250, 152, 72, 53, 74, 201, 123, 251, 44, 78, 45, 96, 34, 161, 170, 74, 204, 67, 41, 92, 161, 102, 11, 242, 113, 100, 253, 20, 12, 104, 1, 2, 49, 0, 160, 241, 212, 158, 153, 228, 100, 10, 187, 83, 190, 131, 195, 187, 101, 143, 24, 111, 155, 226, 205, 101, 66, 117, 32, 223, 110, 20, 255, 45, 243, 172, 22, 45, 199, 5, 18, 195, 121, 78, 168, 138, 131, 210, 98, 203, 45, 72, 89, 2, 56, 48, 130, 2, 52, 48, 130, 1, 186, 160, 3, 2, 1, 2, 2, 16, 86, 37, 83, 149, 199, 167, 251, 64, 235, 226, 40, 216, 38, 8, 83, 182, 48, 10, 6, 8, 42, 134, 72, 206, 61, 4, 3, 3, 48, 75, 49, 31, 48, 29, 6, 3, 85, 4, 3, 12, 22, 65, 112, 112, 108, 101, 32, 87, 101, 98, 65, 117, 116, 104, 110, 32, 82, 111, 111, 116, 32, 67, 65, 49, 19, 48, 17, 6, 3, 85, 4, 10, 12, 10, 65, 112, 112, 108, 101, 32, 73, 110, 99, 46, 49, 19, 48, 17, 6, 3, 85, 4, 8, 12, 10, 67, 97, 108, 105, 102, 111, 114, 110, 105, 97, 48, 30, 23, 13, 50, 48, 48, 51, 49, 56, 49, 56, 51, 56, 48, 49, 90, 23, 13, 51, 48, 48, 51, 49, 51, 48, 48, 48, 48, 48, 48, 90, 48, 72, 49, 28, 48, 26, 6, 3, 85, 4, 3, 12, 19, 65, 112, 112, 108, 101, 32, 87, 101, 98, 65, 117, 116, 104, 110, 32, 67, 65, 32, 49, 49, 19, 48, 17, 6, 3, 85, 4, 10, 12, 10, 65, 112, 112, 108, 101, 32, 73, 110, 99, 46, 49, 19, 48, 17, 6, 3, 85, 4, 8, 12, 10, 67, 97, 108, 105, 102, 111, 114, 110, 105, 97, 48, 118, 48, 16, 6, 7, 42, 134, 72, 206, 61, 2, 1, 6, 5, 43, 129, 4, 0, 34, 3, 98, 0, 4, 131, 46, 135, 47, 38, 20, 145, 129, 2, 37, 185, 245, 252, 214, 187, 99, 120, 181, 245, 95, 63, 203, 4, 91, 199, 53, 153, 52, 117, 253, 84, 144, 68, 223, 155, 254, 25, 33, 23, 101, 198, 154, 29, 218, 5, 11, 56, 212, 80, 131, 64, 26, 67, 79, 178, 77, 17, 45, 86, 195, 225, 207, 191, 203, 152, 145, 254, 192, 105, 96, 129, 190, 249, 108, 188, 119, 200, 141, 221, 175, 70, 165, 174, 225, 221, 81, 91, 90, 250, 171, 147, 190, 156, 11, 38, 145, 163, 102, 48, 100, 48, 18, 6, 3, 85, 29, 19, 1, 1, 255, 4, 8, 48, 6, 1, 1, 255, 2, 1, 0, 48, 31, 6, 3, 85, 29, 35, 4, 24, 48, 22, 128, 20, 38, 215, 100, 217, 197, 120, 194, 90, 103, 209, 167, 222, 107, 18, 208, 27, 99, 241, 198, 215, 48, 29, 6, 3, 85, 29, 14, 4, 22, 4, 20, 235, 174, 130, 196, 255, 161, 172, 91, 81, 212, 207, 36, 97, 5, 0, 190, 99, 189, 119, 136, 48, 14, 6, 3, 85, 29, 15, 1, 1, 255, 4, 4, 3, 2, 1, 6, 48, 10, 6, 8, 42, 134, 72, 206, 61, 4, 3, 3, 3, 104, 0, 48, 101, 2, 49, 0, 221, 139, 26, 52, 129, 165, 250, 217, 219, 180, 231, 101, 123, 132, 30, 20, 76, 39, 183, 91, 135, 106, 65, 134, 194, 177, 71, 87, 80, 51, 114, 39, 239, 229, 84, 69, 126, 246, 72, 149, 12, 99, 46, 92, 72, 62, 112, 193, 2, 48, 44, 138, 96, 68, 220, 32, 31, 207, 229, 155, 195, 77, 41, 48, 193, 72, 120, 81, 217, 96, 237, 106, 117, 241, 235, 74, 202, 190, 56, 205, 37, 184, 151, 208, 200, 5, 190, 240, 199, 247, 139, 7, 165, 113, 198, 232, 14, 7, 104, 97, 117, 116, 104, 68, 97, 116, 97, 88, 152, 236, 10, 37, 162, 251, 198, 240, 34, 22, 0, 231, 241, 54, 234, 147, 135, 17, 159, 246, 180, 35, 47, 6, 136, 101, 107, 116, 231, 32, 141, 47, 231, 69, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 20, 146, 154, 104, 9, 136, 208, 42, 117, 237, 208, 136, 37, 130, 209, 17, 153, 215, 160, 165, 50, 165, 1, 2, 3, 38, 32, 1, 33, 88, 32, 100, 16, 220, 11, 79, 59, 86, 121, 251, 141, 86, 158, 12, 121, 150, 102, 205, 227, 243, 209, 235, 220, 18, 74, 0, 193, 45, 139, 121, 50, 245, 71, 34, 88, 32, 219, 187, 123, 78, 63, 94, 22, 115, 111, 37, 237, 6, 197, 253, 242, 209, 246, 56, 183, 165, 217, 97, 252, 139, 171, 39, 111, 39, 188, 113, 4, 62]), client_data_json: Base64UrlSafeData([123, 34, 116, 121, 112, 101, 34, 58, 34, 119, 101, 98, 97, 117, 116, 104, 110, 46, 99, 114, 101, 97, 116, 101, 34, 44, 34, 99, 104, 97, 108, 108, 101, 110, 103, 101, 34, 58, 34, 70, 71, 84, 122, 115, 45, 86, 112, 82, 80, 72, 108, 116, 51, 106, 70, 86, 84, 77, 56, 45, 54, 108, 112, 104, 73, 85, 57, 76, 99, 109, 86, 56, 107, 121, 67, 119, 121, 106, 71, 110, 97, 99, 34, 44, 34, 111, 114, 105, 103, 105, 110, 34, 58, 34, 104, 116, 116, 112, 115, 58, 47, 47, 97, 109, 101, 116, 104, 121, 115, 116, 46, 108, 111, 99, 97, 108, 58, 56, 52, 52, 51, 34, 125]) }, type_: "public-key" })
```

**Comment by ericmarkmartin** [comment]
Thanks, it works now! I took the `attestation_object` and deserialized it in a python repl with the cbor package and noticed that the format doesn't match the spec; in the spec it seems to indicate that `attStmt` is a key whose value is a map with the single key `x5c`, whereas the deserialized data (below) actually does contain the algorithm.
```
{'fmt': 'apple', 'attStmt': {'alg': -7, 'x5c': [b'0\x82\x02C0\x82\x01\xc9\xa0\x03\x02\x01\x02\x02\x06\x01vER\xfe\xa70\n\x06\x08*\x86H\xce=\x04\x03\x020H1\x1c0\x1a\x06\x03U\x04\x03\x0c\x13Apple WebAuthn CA 11\x130\x11\x06\x03U\x04\n\x0c\nApple Inc.1\x130\x11\x06\x03U\x04\x08\x0c\nCalifornia0\x1e\x17\r201208022715Z\x17\r201211022715Z0\x81\x911I0G\x06\x03U\x04\x03\x0c@9aa90c7c936a4e1bb868965f147a4399f140cf409b434f9059b2d4f5a3cfc0921\x1a0\x18\x06\x03U\x04\x0b\x0c\x11AAA Certification1\x130\x11\x06\x03U\x04\n\x0c\nApple Inc.1\x130\x11\x06\x03U\x04\x08\x0c\nCalifornia0Y0\x13\x06\x07*\x86H\xce=\x02\x01\x06\x08*\x86H\xce=\x03\x01\x07\x03B\x00\x04\xd4\xf8c\x87\xf5N^\xf5\xe7\x16>\xe2-(\xd7\x04\xfb\xbc\xb4}\x16\xec\x85\xa1\xeaN\xfbi\x0bw\x94\x90i\xf9\xc7\xa7\x98\xad^\x939\x02\xfa\x15\x053t\xae\xd9\'\xa0#\x0c\xf9x\xed4\x94\xab\x86\x8a\xcd\x1a\xad\xa3U0S0\x0c\x06\x03U\x1d\x13\x01\x01\xff\x04\x020\x000\x0e\x06\x03U\x1d\x0f\x01\x01\xff\x04\x04\x03\x02\x04\xf003\x06\t*\x86H\x86\xf7cd\x08\x02\x04&0$\xa1"\x04 \xa8\xe2\xa0\xc5=\x92\x0f\xead|\x16\x1d"\x12\xab[\xfdzQ\xf1\xb6i\xf0\xd1\x82\xb0\xb3=T\xb7N\xbe0\n\x06\x08*\x86H\xce=\x04\x03\x02\x03h\x000e\x020\x0e\xf2\x86I\x0c0\x02g\xb8\x84\xbb\x84|\xcc?\x94\xa8N\xe1\xe3\xa1\xf0\x93\xbbZ\xd8A\x9fZjf\xf98\x9c\xc9\xd6\xb6\x0f\xad\xbb\xa7\xf3\x7f\xea\x8a)2>\x021\x00\xc6\x0f\n\xb6\x8egT\x07\x12\x00\xe7\x82\xd6\x1a@:\x11vB\x0e\xc6\xf4:\xd3\x02a\xec\xa3t|I\xa6EEpk\xe4Sh[\xcd\x14\xcb\xfa~\x1d\xbe*', b'0\x82\x0240\x82\x01\xba\xa0\x03\x02\x01\x02\x02\x10V%S\x95\xc7\xa7\xfb@\xeb\xe2(\xd8&\x08S\xb60\n\x06\x08*\x86H\xce=\x04\x03\x030K1\x1f0\x1d\x06\x03U\x04\x03\x0c\x16Apple WebAuthn Root CA1\x130\x11\x06\x03U\x04\n\x0c\nApple Inc.1\x130\x11\x06\x03U\x04\x08\x0c\nCalifornia0\x1e\x17\r200318183801Z\x17\r300313000000Z0H1\x1c0\x1a\x06\x03U\x04\x03\x0c\x13Apple WebAuthn CA 11\x130\x11\x06\x03U\x04\n\x0c\nApple Inc.1\x130\x11\x06\x03U\x04\x08\x0c\nCalifornia0v0\x10\x06\x07*\x86H\xce=\x02\x01\x06\x05+\x81\x04\x00"\x03b\x00\x04\x83.\x87/&\x14\x91\x81\x02%\xb9\xf5\xfc\xd6\xbbcx\xb5\xf5_?\xcb\x04[\xc75\x994u\xfdT\x90D\xdf\x9b\xfe\x19!\x17e\xc6\x9a\x1d\xda\x05\x0b8\xd4P\x83@\x1aCO\xb2M\x11-V\xc3\xe1\xcf\xbf\xcb\x98\x91\xfe\xc0i`\x81\xbe\xf9l\xbcw\xc8\x8d\xdd\xafF\xa5\xae\xe1\xddQ[Z\xfa\xab\x93\xbe\x9c\x0b&\x91\xa3f0d0\x12\x06\x03U\x1d\x13\x01\x01\xff\x04\x080\x06\x01\x01\xff\x02\x01\x000\x1f\x06\x03U\x1d#\x04\x180\x16\x80\x14&\xd7d\xd9\xc5x\xc2Zg\xd1\xa7\xdek\x12\xd0\x1bc\xf1\xc6\xd70\x1d\x06\x03U\x1d\x0e\x04\x16\x04\x14\xeb\xae\x82\xc4\xff\xa1\xac[Q\xd4\xcf$a\x05\x00\xbec\xbdw\x880\x0e\x06\x03U\x1d\x0f\x01\x01\xff\x04\x04\x03\x02\x01\x060\n\x06\x08*\x86H\xce=\x04\x03\x03\x03h\x000e\x021\x00\xdd\x8b\x1a4\x81\xa5\xfa\xd9\xdb\xb4\xe7e{\x84\x1e\x14L\'\xb7[\x87jA\x86\xc2\xb1GWP3r\'\xef\xe5TE~\xf6H\x95\x0cc.\\H>p\xc1\x020,\x8a`D\xdc \x1f\xcf\xe5\x9b\xc3M)0\xc1HxQ\xd9`\xedju\xf1\xebJ\xca\xbe8\xcd%\xb8\x97\xd0\xc8\x05\xbe\xf0\xc7\xf7\x8b\x07\xa5q\xc6\xe8\x0e\x07']}, 'authData': b'\xda\x14\xb1\xf2\xa9\x1e-\xdf\x15-\xfeJ"}\xbc`\x0b\x01G):^\xfc\xb4\xa9\xf3\xd1\x15\xe7\x8a\xb6[E\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x14\xbb\xfbe\x88W\xfeitK\x83\xd5\xc8\xcf\xe4\xaeCE\xc1\x95\xb1\xa5\x01\x02\x03& \x01!X \xd4\xf8c\x87\xf5N^\xf5\xe7\x16>\xe2-(\xd7\x04\xfb\xbc\xb4}\x16\xec\x85\xa1\xeaN\xfbi\x0bw\x94\x90"X i\xf9\xc7\xa7\x98\xad^\x939\x02\xfa\x15\x053t\xae\xd9\'\xa0#\x0c\xf9x\xed4\x94\xab\x86\x8a\xcd\x1a\xad'}
```

**Comment by Firstyear** [comment]
I think reading the spec it does look like attStmt should be a map, so the presence of alg as a key or not shouldn't matter here. I'm curious why alg is there, but it lines up with `COSEContentType::ECDSA_SHA256`. It could be there to indicate the type of cert in x5c or the type of key that is being attested, but certainly I don't see mention of it in the spec. For now, I'd say extract it in the function but note that it is ignored and not used.

**Comment by ericmarkmartin** [comment]
> EDIT: Could be worth checking the latest publish webauthn spec, rather than the draft as it may have changed

The [editor's draft][] doesn't differ in statement format. Looking at @asiegler's comment a bit back and the corresponding implementation it does seem that he is using that parameter to parse `credCert`.

One hiccup I've run into—`TryFrom<(&[u8], COSEContentType)>` for `X509PublicKey` errors out on any ECDSA_SHA256 (the pertinent algo) whose curve is not `NID_X9_62_prime256v1` (Nid value 415, curve `secp256r1`). The touch id attestation, however, appears to use `NID_secp384r1` (Nid value 715, curve `secp384r1`).

I don't know the purpose behind the check, but is the correct approach here to just allow either of the two curves through?


[editor's draft]: https://w3c.github.io/webauthn/#sctn-defined-attestation-formats

**Comment by ericmarkmartin** [comment]
I think we now have a working implementation of apple anonymous attestation verification without steps 4, 5, or the trust chain verification.

**Comment by ericmarkmartin** [comment]
I added a credential registration test helper so one need only provide the registration credential data and re-wrote my test to use it. I'm not sure it saves too much code, but if you like it I can refactor the other registration tests and build another helper for verification. Otherwise, I can just revert it.

**Comment by Firstyear** [comment]
Beside that one comment, this looks like an excellent foundation for touchid attestation. I'd be willing to merge this "as is" before you complete step 4/5. PS: You probably need to rebase your branch to master to account for the changes in src/lib.rs.

**Comment by ericmarkmartin** [comment]
Great! I apologize, which one comment are you referring to?

**Review by Firstyear — COMMENTED** [review]
> Great! I apologize, which one comment are you referring to?
> 
> EDIT: rebased

The comment that apparently I staged and didn't publish because I'm silly. Should be visible now :)

File: src/crypto.rs [hunk 2]
```diff
@@ -99,7 +99,8 @@ impl TryFrom<(&[u8], COSEContentType)> for X509PublicKey {
                     .curve_name()
                     .ok_or(WebauthnError::OpenSSLErrorNoCurveName)?;
 
-                if ec_curve != nid::Nid::X9_62_PRIME256V1 {
+                if ec_curve != nid::Nid::X9_62_PRIME256V1 && ec_curve != nid::Nid::SECP384R1 {
```
[the remark this answers is not in the document]
Comment (ericmarkmartin) [reply src/crypto.rs:102]: It's -7 in my vector, but interestingly it seems your vector doesn't have and alg key.

[the remark this answers is not in the document]
Comment (Firstyear) [reply src/crypto.rs:102]: I wonder if it's to do with the reading that it can be "any curve" but "recommends" SECP256R1 and the -7 just denotes the hashing. Which is why they selected SECP384R1 with sha256, and -7.

[the remark this answers is not in the document]
Comment (Firstyear) [reply src/crypto.rs:102]: Also my test vector isn't the attstmt, it's the full response, you'd need to pass it through the code to get that part out to see if alg is present.

[the remark this answers is not in the document]
Comment (ericmarkmartin) [reply src/crypto.rs:102]: Yeah, I just ran the attestation object through a python script to inspect the data.

[the remark this answers is not in the document]
Comment (ericmarkmartin) [reply src/crypto.rs:102]: I just did another capture and the algorithm is still there.

[the remark this answers is not in the document]
Comment (Firstyear) [reply src/crypto.rs:102]: It's not about the presence of alg or not, it's if the value is -7 or -35, and what that means for accepting different curves in this function. There is a recommendation in the standard that says you *should* use SECP256R1 but it doesn't say you must. It makes me question if apples use of the crypto here is valid (because of the should statement) vs invalid (because it goes against the recommendation). It's not your implementation here that's at fault or at question, I'm wondering about the correctness of the data that apples devices are emitting.

[the remark this answers is not in the document]
Comment (ericmarkmartin) [reply src/crypto.rs:102]: I just did a bit more sleuthing and there's some mention of the `alg` key in a dump someone posts in this PR:
https://github.com/web-auth/webauthn-framework/issues/148. I think it may be from an earlier version of Apple's webauthn format spec and later got dropped. I'm curious, however, as to why my phone continues to send attestation objects with it. For reference I'm on iOS 14.2.

[the remark this answers is not in the document]
Comment (Firstyear) [reply src/crypto.rs:102]: well part of this work is done by the secure enclave, so it could be the model of device in question? What hardware do you have? My challenge was on an ipad mini 5th gen, but I have *not* tried my iphone 7. So I wonder if that's related?

[the remark this answers is not in the document]
Comment (ericmarkmartin) [reply src/crypto.rs:102]: > It's not about the presence of alg or not, it's if the value is -7 or -35, and what that means for accepting different curves in this function. There is a recommendation in the standard that says you _should_ use SECP256R1 but it doesn't say you must. It makes me question if apples use of the crypto here is valid (because of the should statement) vs invalid (because it goes against the recommendation). It's not your implementation here that's at fault or at question, I'm wondering about the correctness of the data that apples devices are emitting.

Right, I understand. I was just trying to confirm that my observation was correct before we went further down the rabbithole. I can confirm that my alg is -7.

> well part of this work is done by the secure enclave, so it could be the model of device in question? What hardware do you have? My challenge was on an ipad mini 5th gen, but I have _not_ tried my iphone 7. So I wonder if that's related?

I'm on an iPhone 8.

**Comment by aseigler** [comment]
> The [editor's draft](https://w3c.github.io/webauthn/#sctn-apple-anonymous-attestation) doesn't differ in statement format. Looking at @aseigler's comment in #33 and the corresponding implementation it does seem that he is using that parameter to parse `credCert`.

To be clear, the alg I was talking about is the one inside the credential public key, within attested credential data, within authData, *NOT* within attStmt..

Diagram

**Comment by ericmarkmartin** [comment]
Sorry, that's my bad for misreading. I saw that you observed a -7 as well and jumped the gun a bit.

Did you ever observe an alg key in the attStmt map?

**Comment by aseigler** [comment]
> Sorry, that's my bad for misreading. I saw that you observed a -7 as well and jumped the gun a bit.
> 
> Did you ever observe an alg key in the attStmt map?

Yes, see here:  https://bugs.webkit.org/show_bug.cgi?id=217720

**Comment by ericmarkmartin** [comment]
Ah, many thanks

**Comment by ericmarkmartin** [comment]
Use of the alg field has been reverted. I'm hardcoding in ECDSA_SHA384 as the COSE content type for now.

File: src/crypto.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ericmarkmartin) [reply src/crypto.rs]: See conversation below---this field has been removed from newer versions of webkit (although it's strange I'm still seeing it).

**Comment by Firstyear** [comment]
Yep seems good to me :)

**Comment by ericmarkmartin** [comment]
If you want to merge I can either rebase and continue on this branch or make a new branch to tackle 4 and 5

**Comment by Firstyear** [comment]
Can you rebase and squash to a single commit, then I'll merge this, then you can do 4/5 in a new branch?

**Comment by ericmarkmartin** [comment]
Sure--can you not squash the commits in the merge?

**Comment by Firstyear** [comment]
Apparently I'm half asleep today :)

**Comment by aseigler** [comment]
> Use of the alg field has been reverted. I'm hardcoding in ECDSA_SHA384 as the COSE content type for now.

FWIW you should be able to use credential_public_key.type_ instead of hard coding ECDSA_SHA384.
